src/xtgeo/surface/_regsurf_gridding.py

```diff
@@ -322,21 +322,26 @@ def surf_fill(self, fill_value=None):
         logger.info("Do fill... DONE")
 
 
-def smooth_median(self, iterations=1, width=1):
-    """Smooth a surface using a median filter.
+def _smooth(
+    self,
+    window_function: Callable[[np.ndarray], np.ndarray],
+    iterations: int = 1,
+    width: int = 1,
+) -> None:
+    """
+    Smooth a RegularSurface using a window function.
 
-    .. versionadded:: 2.1
+    Original mask (undefined values) is stored before applying
+    smoothing on a filled array. Subsequently the original mask
+    is used to restore the undefined values in the output.
     """
 
     mask = ma.getmaskarray(self.values)
-    tmpv = ma.filled(self.values, fill_value=np.nan)
 
-    for _itr in range(iterations):
-        tmpv = scipy.ndimage.median_filter(tmpv, width)
+    self.fill()
 
-    tmpv = ma.masked_invalid(tmpv)
+    smoothed_values = self.values
```

Comment:
The smoothed_values now is pointing to the same object as self.values. I think making a copy here is safer?

`smoothed_values = self.values.copy()`

Reply:
I had this thought as well but I guess we should prefer to work on large arrays in-place when possible to avoid the time and memory cost of copying. Maybe @tnatt and @janbjorge had this discussion; thoughts? I'm still not super solid with numpy wrt these details.

Reply:
Yes it may work, but my experience is that you need to consider at least:
```
    for _ in range(iterations):
        smoothed_values = window_function(smoothed_values, width, mode="nearest")
```
For the first loop, the `smoothed_values` in `window_function()` will be `self.values`. If the `window_function` always returns a copy, it will be safe; if it return a view; then I think `self.values` is modified here in the loop?

Reply:
Since self.values has getters and setters that has quite a bit of logic attached i believe i it will be a lot quicker to work on a copy instead of re-doing the "checks" for every set in for-loop.

Reply:
I must admit, I am not sure what is the best approach here.. 
Initially I wrote the code without using an extra `smoothed_values `symbol, just directly wrote `self.values = window_function(self.values, ..)` since it is safe to modify it. But JB pointed at that extra overhead time (due to the getters and setters of this property) we might get by doing that. So we landed on having it like this, but then I probably should have made it a copy as Jan mentions 🙂

Reply:
Tested it now, and saw that the current implementation does not invoke the getter/setter of the self.values 👍

Reply:
Hmmm It should, but the _getter_ is only returning self._values (the np array) so there should be no overhead, For the _setter_ there is some overhead with sanity checks

Reply:
yes my mistake does not invoke the setter I meant :)